tests/heka/test_message_parser.py

```diff
@@ -1,87 +1,105 @@
+    ("mixed", False, False, 5, False),
+    ("gzip",  False, False, 10, False),
+    ("gzip_mixed", False, False, 5, False),
+    # snappy enabled
+    ("plain", True, False, 10, False),
+    ("snappy", True, False, 10, False),
+    ("mixed", True, False, 10, False),
+    ("gzip",  True, False, 10, False),
+    ("gzip_mixed", True, False, 10, False),
+    # strict mode on
+    ("plain", False, True, 10, False),
+    ("snappy", False, True, 0, True),
+    ("mixed", False, True, 5, True),
+    ("gzip",  False, True, 10, False),
+    ("gzip_mixed", False, True, 5, True)
+])
+def test_unpack(data_dir, heka_format, try_snappy, strict, expected_count,
+                expected_exception):
+    count = 0
+    threw_exception = False
+    filename = "{}/test_{}.heka".format(data_dir, heka_format)
+    with open(filename, "rb") as o:
+        if "gzip" in heka_format:
+            o = streaming_gzip_wrapper(o)
+        try:
+            for r, b in message_parser.unpack(o, try_snappy=try_snappy, strict=strict):
                 j = json.loads(r.message.payload)
-                assert msg == j["seq"]
-                msg += 1
-            assert 10 == msg
+                assert count == j["seq"]
+                count += 1
+        except DecodeError:
+            threw_exception = True
 
+    assert count == expected_count
+    assert threw_exception == expected_exception
 
-def test_unpack_nosnappy(data_dir):
-    expected_counts = {"plain": 10, "snappy": 0, "mixed": 5,
-                       "gzip": 10, "gzip_mixed": 5}
-    for t in expected_counts.keys():
-        count = 0
-        filename = "{}/test_{}.heka".format(data_dir, t)
-        with open(filename, "rb") as o:
-            if "gzip" in t:
-                o = streaming_gzip_wrapper(o)
-            try:
-                for r, b in message_parser.unpack(o, try_snappy=False):
-                    count += 1
-            except:
-                pass
-            assert expected_counts[t] == count
 
+@pytest.mark.parametrize("heka_format", ["snappy", "gzip"])
+def test_telemetry(data_dir, heka_format):
+    filename = "{}/test_telemetry_{}.heka".format(data_dir, heka_format)
+    reference_filename = filename + '.json'
 
-def test_unpack_strict(data_dir):
-    expected_exceptions = {"plain": False, "snappy": True, "mixed": True,
-                           "gzip": False, "gzip_mixed": True}
-    for t in expected_exceptions.keys():
-        count = 0
-        filename = "{}/test_{}.heka".format(data_dir, t)
-        threw = False
-        with open(filename, "rb") as o:
-            if "gzip" in t:
-                o = streaming_gzip_wrapper(o)
-            try:
-                for r, b in message_parser.unpack(o, strict=True, try_snappy=False):
-                    count += 1
-            except Exception:
-                threw = True
-        assert expected_exceptions[t] == threw
+    # enable this to regenerate the expected json representation of the ping
```

Comment:
I'm not sure I like adding the generated JSON reference blobs to the repo, as the real reference blobs are those submitted by the client. I think the better thing to do would be to use a raw ping from the client (from e.g. about:telemetry) as the reference, and re-derive the legacy heka decoded and current infra decoded versions from that. This it turns out is not trivial to do, esp. for the deprecated infra, but it is possible. The test as written assumes that our code is working correctly now, so as a regression test it is acceptable. I don't think it's necessarily worth the effort involved to recreate our test data using an actual reference ping, but I thought I would call it out.